patchwork/views/patch.py

```diff
@@ -64,31 +68,88 @@ def patch_detail(request, project_id, msgid):
 
     form = None
     create_bundle_form = None
+    create_note_form = None
+    edit_note_form = None
+    edit_note_id = None
+    edit_cancel = False
     errors = None
+    is_maintainer = (
+        request.user.is_superuser
+        or request.user.is_authenticated
+        and patch.project in request.user.profile.maintainer_projects.all()
+    )
 
     if editable:
         form = PatchForm(instance=patch)
     if request.user.is_authenticated:
         create_bundle_form = CreateBundleForm()
 
     if request.method == 'POST':
-        action = request.POST.get('action', None)
-        if action:
-            action = action.lower()
-
-        if action in ['create', 'add']:
-            errors = set_bundle(
-                request, project, action, request.POST, [patch]
+        form_name = request.POST.get('form_name', '')
+        if is_maintainer and form_name == DeleteNoteForm.name:
+            note_form = DeleteNoteForm(request.POST)
+            if note_form.is_valid():
+                note = get_object_or_404(
+                    Note, id=note_form.cleaned_data['note_id']
+                )
+                note.delete()
+                messages.success(request, 'Note removed')
+            else:
+                errors = note_form.errors
+
+        elif is_maintainer and form_name == CreateNoteForm.name:
+            create_note = CreateNoteForm(
+                request.POST,
+                instance=Note(patch=patch, submitter=request.user),
             )
+            if create_note.is_valid():
+                create_note.save()
+                messages.success(request, 'Note created')
+            create_note_form = None
+
+        elif is_maintainer and form_name == EditNoteForm.name:
+            edit_cancel = not not request.POST.get('cancel', None)
+            if not edit_cancel:
+                edit_note_form = EditNoteForm(request.POST)
+                if edit_note_form.is_valid():
+                    note = get_object_or_404(
+                        Note, id=request.POST.get('note_id'), patch=patch
+                    )
+                    edit_note_form = EditNoteForm(request.POST, instance=note)
+                    if edit_note_form.is_valid():
+                        edit_note_form.save()
+                        messages.success(request, 'Note updated')
+                else:
+                    errors = edit_note_form.err
+        else:
+            action = request.POST.get('action', None)
+            if action:
+                action = action.lower()
+
+            if is_maintainer and action == 'add-note':
+                create_note_form = CreateNoteForm()
+
+            if is_maintainer and action == 'edit-note':
```

Comment:
You have to make sure the other actions are not evaluated after the first match, without the elif when trying to add a note as superuser but not maintainer we get a 403 error.
```suggestion
            elif is_maintainer and action == 'edit-note':
```